syntax = "proto2";

package SimpleChat;

enum UserPresence {
    ONLINE = 0;
    AWAY = 1;
    BUSY = 2;
}

enum UserStatus {
    NONE = 0;
    MUTED = 1;
}

enum UserAction {
    JOINED = 0;
    LEFT = 1;
    KICKED = 2;
}

message User {
    required int32 id = 1;
    required string name = 2;
    required UserPresence presence = 3 [default = ONLINE];
    required UserStatus status = 4 [default = NONE];
}

message UserJoinRequest {
    required string name = 1;
}

message UserJoinResponse {
    required bool success = 1;

    optional User user = 2;
    optional string motd = 3;
    optional string message = 4;
}

message UserListRequest {
    required string name = 1;
}

message UserListResponse {
    repeated User users = 1;
    optional string message = 2;
}

message UserChange {
    required User user = 1;
    optional UserPresence presence = 2;
    optional UserStatus status = 3;
    optional UserAction action = 4;
}